### DISCUSSION TITLE
Treat double precision floating-point constant as single precision constant.

### DISCUSSION THREAD

**Discussion by zjin-lcf** [intel/llvm#3806] [General]
Does the DPCPP compiler have an option  for a single-source SYCL program ?  Thanks.

  -cl-single-precision-constant
                          OpenCL only. Treat double precision floating-point constant as single precision constant.

**Comment by bader** [comment]
As it's specified in the description `-cl-single-precision-constant` is an OpenCL option, so I would expect DPC++ compiler to ignore this option for SYCL mode.
You can try to specify this option to DPC++ back-end compiler (which might be a SPIR-V compiler supporting this option), but depending on the use case it might be too late (e.g. if constant type influenced type deduction, overload resolution, etc.).

Could you share the use case(s) where this option is useful, please?

**Reply by zjin-lcf** [reply]
A use case is SHOC's s3d. The benchmark contains many double-precision floating-point operations. Please see the updates.
https://github.com/zjin-lcf/oneAPI-DirectProgramming/commit/1431591b33aa579f7ae3aa3e0e8fea95d4ea7108

**Reply by bader** [reply]
> You can try to specify this option to DPC++ back-end compiler (which might be a SPIR-V compiler supporting this option), but depending on the use case it might be too late (e.g. if constant type influenced type deduction, overload resolution, etc.).

This approach won't work for the code like this (assuming math functions are not evaluated by early optimizer):
```C++
RKLOW(1) = EXP(4.22794408e1 -9.e-1*ALOGT + DIV(8.55468335e2,TEMP));
```

Technically the applications must check if `sycl::aspect::fp64` is supported by the device before using such constants, but I see how an option similar to `-cl-single-precision-constant` can simplify such code.

AFAIK, DPC++ doesn't support such option, so feel free to file a feature request.
